### ISSUE TITLE
Improve the lepton library

### ISSUE THREAD

**Issue by HanatoK** [Colvars/colvars#231]
It would be useful to make the custom function support functions like atan2. The lepton used in PLUMED adds a bunch of trigonometric functions, and I swipe the code there and adapt a patch for NAMD. Can the improvement of lepton also be applied to VMD or LAMMPS?

```
diff --git a/lepton/include/lepton/Operation.h b/lepton/include/lepton/Operation.h
index f7a8b781..33bf7306 100644
--- a/lepton/include/lepton/Operation.h
+++ b/lepton/include/lepton/Operation.h
@@ -64,7 +64,8 @@ public:
      */
     enum Id {CONSTANT, VARIABLE, CUSTOM, ADD, SUBTRACT, MULTIPLY, DIVIDE, POWER, NEGATE, SQRT, EXP, LOG,
              SIN, COS, SEC, CSC, TAN, COT, ASIN, ACOS, ATAN, SINH, COSH, TANH, ERF, ERFC, STEP, DELTA, SQUARE, CUBE, RECIPROCAL,
-             ADD_CONSTANT, MULTIPLY_CONSTANT, POWER_CONSTANT, MIN, MAX, ABS, FLOOR, CEIL, SELECT};
+             ADD_CONSTANT, MULTIPLY_CONSTANT, POWER_CONSTANT, MIN, MAX, ABS, FLOOR, CEIL, SELECT,
+             ACOT, ASEC, ACSC, COTH, SECH, CSCH, ASINH, ACOSH, ATANH, ACOTH, ASECH, ACSCH, ATAN2};
     /**
      * Get the name of this Operation.
      */
@@ -156,6 +157,19 @@ public:
     class Floor;
     class Ceil;
     class Select;
+    class Acot;
+    class Asec;
+    class Acsc;
+    class Coth;
+    class Sech;
+    class Csch;
+    class Asinh;
+    class Acosh;
+    class Atanh;
+    class Acoth;
+    class Asech;
+    class Acsch;
+    class Atan2;
 };
 
 class LEPTON_EXPORT Operation::Constant : public Operation {
@@ -834,6 +848,7 @@ public:
     ExpressionTreeNode differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const;
 };
 
+
 class LEPTON_EXPORT Operation::Square : public Operation {
 public:
     Square() {
@@ -1160,6 +1175,46 @@ public:
     ExpressionTreeNode differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const;
 };
 
-} // namespace Lepton
+#define LEPTON_CLASS_OPERATION(Name,name,NAME,nargs,impl) \
+class LEPTON_EXPORT Operation::Name : public Operation { \
+public: \
+    Name() { \
+    } \
+    std::string getName() const { \
+        return #name; \
+    } \
+    Id getId() const { \
+        return NAME; \
+    } \
+    int getNumArguments() const { \
+        return nargs; \
+    } \
+    Operation* clone() const { \
+        return new Name(); \
+    } \
+    double evaluate(double* args, const std::map<std::string, double>& variables) const { \
+        return impl; \
+    } \
+    ExpressionTreeNode differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const; \
+}
+
+LEPTON_CLASS_OPERATION(Acot,acot,ACOT,1,std::atan(1.0/args[0]));
+LEPTON_CLASS_OPERATION(Asec,asec,ASEC,1,std::acos(1.0/args[0]));
+LEPTON_CLASS_OPERATION(Acsc,acsc,ACSC,1,std::asin(1.0/args[0]));
+LEPTON_CLASS_OPERATION(Coth,coth,ACOT,1,1.0/std::tanh(args[0]));
+LEPTON_CLASS_OPERATION(Sech,sech,SECH,1,1.0/std::cosh(args[0]));
+LEPTON_CLASS_OPERATION(Csch,csch,CSCH,1,1.0/std::sinh(args[0]));
+
+LEPTON_CLASS_OPERATION(Asinh,asinh,ASINH,1,std::asinh(args[0]));
+LEPTON_CLASS_OPERATION(Acosh,acosh,ACOSH,1,std::acosh(args[0]));
+LEPTON_CLASS_OPERATION(Atanh,atanh,ATANH,1,std::atanh(args[0]));
+
+LEPTON_CLASS_OPERATION(Acoth,acoth,ACOTH,1,0.5*std::log((args[0]+1.0)/(args[0]-1.0)));
+LEPTON_CLASS_OPERATION(Asech,asech,ASECH,1,std::log(std::sqrt(1.0/args[0]-1.0)*std::sqrt(1.0/args[0]+1.0)+1.0/args[0]));
+LEPTON_CLASS_OPERATION(Acsch,acsch,ACSCH,1,std::log(1.0/args[0]+std::sqrt(1.0/(args[0]*args[0])+1.0)));
+
+LEPTON_CLASS_OPERATION(Atan2,atan2,ATAN2,2,std::atan2(args[0],args[1]));
+
+} // namespace lepton
 
 #endif /*LEPTON_OPERATION_H_*/
diff --git a/lepton/src/Operation.cpp b/lepton/src/Operation.cpp
index 693dea2e..fdafc698 100644
--- a/lepton/src/Operation.cpp
+++ b/lepton/src/Operation.cpp
@@ -1,4 +1,3 @@
-
 /* -------------------------------------------------------------------------- *
  *                                   Lepton                                   *
  * -------------------------------------------------------------------------- *
@@ -34,6 +33,10 @@
 #include "lepton/ExpressionTreeNode.h"
 #include "MSVC_erfc.h"
 
+#ifndef M_PI
+#define M_PI           3.14159265358979323846
+#endif
+
 using namespace Lepton;
 using namespace std;
 
@@ -333,3 +336,219 @@ ExpressionTreeNode Operation::Select::differentiate(const std::vector<Expression
     derivChildren.push_back(childDerivs[2]);
     return ExpressionTreeNode(new Operation::Select(), derivChildren);
 }
+
+#define LEPTON_CONST(x) ExpressionTreeNode(new Operation::Constant(x))
+#define LEPTON_OP1(name,x) ExpressionTreeNode(new Operation::name(),x)
+#define LEPTON_OP2(name,x,y) ExpressionTreeNode(new Operation::name(),x,y)
+#define LEPTON_ADD_CONST(x,y) ExpressionTreeNode(new Operation::AddConstant(x),y)
+
+ExpressionTreeNode Operation::Acot::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+      LEPTON_OP2(Multiply,
+        LEPTON_OP1(Negate,
+          LEPTON_OP1(Reciprocal,
+            LEPTON_ADD_CONST(1.0,
+              LEPTON_OP1(Square,children[0])
+            )
+          )
+        ),
+        childDerivs[0]
+      );
+}
+
+ExpressionTreeNode Operation::Asec::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+      LEPTON_OP2(Multiply,
+        LEPTON_OP1(Reciprocal,
+          LEPTON_OP2(Multiply,
+            LEPTON_OP1(Abs,children[0]),
+            LEPTON_OP1(Sqrt,
+              LEPTON_OP2(Subtract,
+                LEPTON_OP1(Square,children[0]),
+                LEPTON_CONST(1.0)
+              )
+            )
+          )
+        ),
+        childDerivs[0]
+      );
+}
+
+ExpressionTreeNode Operation::Acsc::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Negate,
+        LEPTON_OP1(Reciprocal,
+          LEPTON_OP2(Multiply,
+            LEPTON_OP1(Abs,children[0]),
+            LEPTON_OP1(Sqrt,
+              LEPTON_OP2(Subtract,
+                LEPTON_OP1(Square,children[0]),
+                LEPTON_CONST(1.0)
+              )
+            )
+          )
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Coth::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP2(Subtract,
+        LEPTON_CONST(1.0),
+        LEPTON_OP1(Square,
+          LEPTON_OP1(Coth,children[0])
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Sech::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Negate,
+        LEPTON_OP2(Multiply,
+          LEPTON_OP1(Tanh,children[0]),
+          LEPTON_OP1(Sech,children[0])
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Csch::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Negate,
+        LEPTON_OP2(Multiply,
+          LEPTON_OP1(Coth,children[0]),
+          LEPTON_OP1(Csch,children[0])
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Acosh::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Reciprocal,
+        LEPTON_OP1(Sqrt,
+          LEPTON_OP2(Subtract,
+            LEPTON_OP1(Square,children[0]),
+            LEPTON_CONST(1.0)
+          )
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Atanh::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Reciprocal,
+        LEPTON_OP2(Subtract,
+          LEPTON_CONST(1.0),
+          LEPTON_OP1(Square,children[0])
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Asinh::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Reciprocal,
+        LEPTON_OP1(Sqrt,
+          LEPTON_ADD_CONST(1.0,
+            LEPTON_OP1(Square,children[0])
+          )
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Acoth::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Reciprocal,
+        LEPTON_OP2(Subtract,
+          LEPTON_CONST(1.0),
+          LEPTON_OP1(Square,children[0])
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Asech::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Negate,
+        LEPTON_OP1(Reciprocal,
+          LEPTON_OP2(Multiply,
+            children[0],
+            LEPTON_OP2(Multiply,
+              LEPTON_ADD_CONST(1.0,
+                children[0]
+              ),
+              LEPTON_OP1(Sqrt,
+                LEPTON_OP2(Divide,
+                  LEPTON_OP2(Subtract,
+                    LEPTON_CONST(1.0),
+                    children[0]
+                  ),
+                  LEPTON_ADD_CONST(1.0,
+                    children[0]
+                  )
+                )
+              )
+            )
+          )
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Acsch::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Multiply,
+      LEPTON_OP1(Negate,
+        LEPTON_OP1(Reciprocal,
+          LEPTON_OP2(Multiply,
+            LEPTON_OP1(Square,children[0]),
+            LEPTON_OP1(Sqrt,
+              LEPTON_ADD_CONST(1.0,
+                LEPTON_OP1(Reciprocal,
+                  LEPTON_OP1(Square,children[0])
+                )
+              )
+            )
+          )
+        )
+      ),
+      childDerivs[0]
+    );
+}
+
+ExpressionTreeNode Operation::Atan2::differentiate(const std::vector<ExpressionTreeNode>& children, const std::vector<ExpressionTreeNode>& childDerivs, const std::string& variable) const {
+    return
+    LEPTON_OP2(Divide,
+      LEPTON_OP2(Subtract,
+        LEPTON_OP2(Multiply, children[1], childDerivs[0]),
+        LEPTON_OP2(Multiply, children[0], childDerivs[1])
+      ),
+      LEPTON_OP2(Add,
+        LEPTON_OP1(Square, children[0]),
+        LEPTON_OP1(Square, children[1])
+      )
+    );
+}
diff --git a/lepton/src/Parser.cpp b/lepton/src/Parser.cpp
index 6b19d737..bc951e49 100644
--- a/lepton/src/Parser.cpp
+++ b/lepton/src/Parser.cpp
@@ -41,12 +41,17 @@
 using namespace Lepton;
 using namespace std;
 
+namespace Lepton {
+
 static const string Digits = "0123456789";
 static const string Operators = "+-*/^";
 static const bool LeftAssociative[] = {true, true, true, true, false};
 static const int Precedence[] = {0, 0, 1, 1, 3};
 static const Operation::Id OperationId[] = {Operation::ADD, Operation::SUBTRACT, Operation::MULTIPLY, Operation::DIVIDE, Operation::POWER};
 
+}
+
+
 class Lepton::ParseToken {
 public:
     enum Type {Number, Operator, Variable, Function, LeftParen, RightParen, Comma, Whitespace};
@@ -299,37 +304,49 @@ Operation* Parser::getOperatorOperation(const std::string& name) {
 
 Operation* Parser::getFunctionOperation(const std::string& name, const map<string, CustomFunction*>& customFunctions) {
 
-    static map<string, Operation::Id> opMap;
-    if (opMap.size() == 0) {
-        opMap["sqrt"] = Operation::SQRT;
-        opMap["exp"] = Operation::EXP;
-        opMap["log"] = Operation::LOG;
-        opMap["sin"] = Operation::SIN;
-        opMap["cos"] = Operation::COS;
-        opMap["sec"] = Operation::SEC;
-        opMap["csc"] = Operation::CSC;
-        opMap["tan"] = Operation::TAN;
-        opMap["cot"] = Operation::COT;
-        opMap["asin"] = Operation::ASIN;
-        opMap["acos"] = Operation::ACOS;
-        opMap["atan"] = Operation::ATAN;
-        opMap["sinh"] = Operation::SINH;
-        opMap["cosh"] = Operation::COSH;
-        opMap["tanh"] = Operation::TANH;
-        opMap["erf"] = Operation::ERF;
-        opMap["erfc"] = Operation::ERFC;
-        opMap["step"] = Operation::STEP;
-        opMap["delta"] = Operation::DELTA;
-        opMap["square"] = Operation::SQUARE;
-        opMap["cube"] = Operation::CUBE;
-        opMap["recip"] = Operation::RECIPROCAL;
-        opMap["min"] = Operation::MIN;
-        opMap["max"] = Operation::MAX;
-        opMap["abs"] = Operation::ABS;
-        opMap["floor"] = Operation::FLOOR;
-        opMap["ceil"] = Operation::CEIL;
-        opMap["select"] = Operation::SELECT;
-    }
+    const static map<string, Operation::Id> opMap ={
+        { "sqrt" , Operation::SQRT },
+        { "exp" , Operation::EXP },
+        { "log" , Operation::LOG },
+        { "sin" , Operation::SIN },
+        { "cos" , Operation::COS },
+        { "sec" , Operation::SEC },
+        { "csc" , Operation::CSC },
+        { "tan" , Operation::TAN },
+        { "cot" , Operation::COT },
+        { "asin" , Operation::ASIN },
+        { "acos" , Operation::ACOS },
+        { "atan" , Operation::ATAN },
+        { "sinh" , Operation::SINH },
+        { "cosh" , Operation::COSH },
+        { "tanh" , Operation::TANH },
+        { "erf" , Operation::ERF },
+        { "erfc" , Operation::ERFC },
+        { "step" , Operation::STEP },
+        { "delta" , Operation::DELTA },
+        { "square" , Operation::SQUARE },
+        { "cube", Operation::CUBE },
+        { "recip" , Operation::RECIPROCAL },
+        { "min" , Operation::MIN },
+        { "max" , Operation::MAX },
+        { "abs" , Operation::ABS },
+        { "floor" , Operation::FLOOR },
+        { "ceil" , Operation::CEIL },
+        { "select" , Operation::SELECT },
+        { "acot" , Operation::ACOT },
+        { "asec" , Operation::ASEC },
+        { "acsc" , Operation::ACSC },
+        { "coth" , Operation::COTH },
+        { "sech" , Operation::SECH },
+        { "csch" , Operation::CSCH },
+        { "asinh" , Operation::ASINH },
+        { "acosh" , Operation::ACOSH },
+        { "atanh" , Operation::ATANH },
+        { "acoth" , Operation::ACOTH },
+        { "asech" , Operation::ASECH },
+        { "acsch" , Operation::ACSCH },
+        { "atan2" , Operation::ATAN2 },
+    };
     string trimmed = name.substr(0, name.size()-1);
 
     // First check custom functions.
@@ -400,6 +417,32 @@ Operation* Parser::getFunctionOperation(const std::string& name, const map<strin
             return new Operation::Ceil();
         case Operation::SELECT:
             return new Operation::Select();
+        case Operation::ACOT:
+            return new Operation::Acot();
+        case Operation::ASEC:
+            return new Operation::Asec();
+        case Operation::ACSC:
+            return new Operation::Acsc();
+        case Operation::COTH:
+            return new Operation::Coth();
+        case Operation::SECH:
+            return new Operation::Sech();
+        case Operation::CSCH:
+            return new Operation::Csch();
+        case Operation::ASINH:
+            return new Operation::Asinh();
+        case Operation::ACOSH:
+            return new Operation::Acosh();
+        case Operation::ATANH:
+            return new Operation::Atanh();
+        case Operation::ACOTH:
+            return new Operation::Acoth();
+        case Operation::ASECH:
+            return new Operation::Asech();
+        case Operation::ACSCH:
+            return new Operation::Acsch();
+        case Operation::ATAN2:
+            return new Operation::Atan2();
         default:
             throw Exception("unknown function");
     }
```

**Comment by giacomofiorin** [comment]
Lepton is available in LAMMPS out of the box.  However, it has so far been embargoed by the VMD maintainers: it would be great if you could advocate for it.

Regarding modifying Lepton, it is a good idea but it would be appropriate to modify it at the source. 

@GiovanniBussi have you contacted @peastman regarding having a standard repository for Lepton?

**Comment by peastman** [comment]
Lepton is developed as part of OpenMM, so the "canonical" version is always found in https://github.com/pandegroup/openmm/ under libraries/lepton.  I'm happy to include additional trig functions, if there's demand for that, or you can just implement them with CustomFunction.

**Comment by HanatoK** [comment]
@peastman I have a quick look at the trigonometric functions in the patch, and it seems only `atan2(y, x)` cannot directly implement by CustomFunction. Can you just include the `atan2` function in Lepton?

**Comment by GiovanniBussi** [comment]
Hi all,

since we embedded Lepton in PLUMED our implementation diverged slightly. You can see all changes [here](https://github.com/plumed/plumed2/commits/master/src/lepton). In case some of them is interesting for the community, I am happy to make a patch for the canonical openmm version. The patch @HanatoK did seems perfect for the trigonometric functions.

About using CustomFunction for trigonometric functions, I am not sure I remember correctly but I think that it requires you to specify also higher order derivatives. To me it seemed simpler to just define a few additional functions following those already present.

I mostly added functions in order to improve compatibility with libmatheval, since as of PLUMED 2.5 we don't link libmatheval anymore and use Lepton for everything. I also changed the definition of some derivatives in limiting cases (e.g., in our version the derivative of a step function is a delta and the derivative of a delta is a nandelta, consistently with libmatheval).

@peastman I also updated our embedded Lepton to the latest asmjit. If you are interested I can check if this can be backported to openmm (this would require also the asmjit version embedded in openmm to be updated).

Thanks!

Giovanni

**Comment by peastman** [comment]
> it seems only atan2(y, x) cannot directly implement by CustomFunction.

Why not?

> I also updated our embedded Lepton to the latest asmjit.

I've been meaning to update that for a long time.  This would definitely be good to do.

**Comment by GiovanniBussi** [comment]
@peastman great I will try to port the upgraded asmjit to openmm and open a pull request to openmm repository. Thanks! Giovanni

**Comment by peastman** [comment]
Thanks!  I've actually just been working on that, but haven't managed to get it working yet.  A lot seems to have changed.  So if you already have it working, that's great.

**Comment by GiovanniBussi** [comment]
I think I have it already - I am able to compile openmm with the updated asmjit/lepton on my laptop. I just ported the changes from the plumed tree. Is there a easy way to test it in openmm? Tomorrow I’ll push it on my github fork and see if it passes Travis tests

**Comment by peastman** [comment]
Run the `TestParser` test case.  If anything's wrong, that will probably catch it.

**Comment by HanatoK** [comment]
> > it seems only atan2(y, x) cannot directly implement by CustomFunction.
> 
> Why not?

I misinterpreted what the CustomFunction refers. In colvars it's a keyword to pass a math expression to lepton and the users cannot use existing functions to emulate atan2. However if you refer it to the CustomFunction class in the C++ code, it definitely can be extended with atan2.
Thanks to everyone for improving the lepton library!

**Comment by jhenin** [comment]
Great. We will pull from upstream when the changes get there.

**Comment by giacomofiorin** [comment]
@GiovanniBussi As mentioned, Lepton 1.3 is already in LAMMPS within the `lib/colvars` folder. This avoided having to create a "USER-" package just for it. The vast majority of architectures used with LAMMPS support Colvars as well, and Lepton is included this way every time.

If you ever find a use case where this arrangement doesn't work, feel free to submit a change to move Lepton outside the `lib/colvars` folder into a new location (e.g. `lib/lepton`).

**Comment by peastman** [comment]
@GiovanniBussi are you still planning to send a PR with this, or should I try to get my version working?

**Comment by GiovanniBussi** [comment]
Hi @peastman sorry for lagging on this...

I managed to have a version that seems to work [here](https://github.com/GiovanniBussi/openmm/commits/upgrade-asmjit), at least on my laptop.

However, I am breaking some of the regtests on travis. I tried also to push intermediate commits to see if they work (https://travis-ci.org/GiovanniBussi/openmm/builds).

Maybe for you it is easier to tell if failures are expected or not?

Thanks!

Giovanni
